### PR TITLE
Explore: fixes toolbars datasource selector and date picker responsiveness

### LINKED ISSUE #19351: Explore: Toolbar's date picker not responsive enough with long dates
This toolbar is getting busier and looks messy in split views:

Screenshot 2019-09-24 at 10 40 41

### LINKED ISSUE #19353: Explore: Toolbar's datasource selector not responsive
This toolbar is getting busier and looks messy in split views:

Screenshot 2019-09-24 at 10 40 41

The picker could just show the icon.

### PR DESCRIPTION
**Pull request by Estrax**
**What this PR does / why we need it**:

Adds responsive styles for datasource selector and updates responsiveness of date picker with long dates.

**Which issue(s) this PR fixes**:

Fixes #19353 
Fixes #19351

### CODE REVIEW

**Comment by Estrax** [comment]
@kaydelaney @davkal wdyt?

**Review by davkal — DISMISSED** [review]
Works well and could go in as is.

File: packages/grafana-ui/src/components/TimePicker/TimePicker.tsx [hunk 1]
```diff
@@ -183,6 +183,7 @@ class UnThemedTimePicker extends PureComponent<Props, State> {
       </>
     );
     const isAbsolute = isDateTime(value.raw.to);
+    const isLongDate = typeof this.props.value.raw.from !== 'string' || typeof this.props.value.raw.to !== 'string';
```
Comment (davkal) [inline packages/grafana-ui/src/components/TimePicker/TimePicker.tsx]: Wouldnt `isAbsolute` be enough?

**Comment by davkal** [comment]
@aocenas would that have been a good use for the `ResponsiveButton`?

**Comment by aocenas** [comment]
Should the datasource picker also be responsive? At the moment this is what I see on tablet size:
Screenshot 2019-10-09 at 10 23 07

> would that have been a good use for the ResponsiveButton?

Probably yes. Though not sure how TimePicker is done atm and whether it will be easy to just inject it there. Also as I think TimePicker is one that is used everywhere else it will make it responsive everywhere.

As a side note, I do not particularly like the approach of using just the icons. For things like run button it would make sense as the `Run` text does not change but with TimePicker you loose the info of what range we are showing and I doubt people know datasources by icons. Personally I would rather have it just stacked so it does not overflow, which is uglier but more functional and maybe use some more compact date format.

File: packages/grafana-ui/src/components/TimePicker/TimePicker.tsx (hunk 1 above)
Comment (Estrax) [reply packages/grafana-ui/src/components/TimePicker/TimePicker.tsx]: @davkal Sure, though we are going to have small inconsistencies then.

> 2019-10-01 00:00:00 to a few seconds ago
aka **date-to-now** would be a short date (with isAbsolute).

> 5 minutes ago to 2019-10-09 23:59:59
aka **now-minus-offset-to-date** would be a long date, even though technically it's shorter than date-to-now.

With **isLongDate** both of them are classified as long dates.

Comment (Estrax) [reply packages/grafana-ui/src/components/TimePicker/TimePicker.tsx]: @davkal Sure, though we are going to have small inconsistencies then.

With isAbsolute:

> 2019-10-01 00:00:00 to a few seconds ago

aka **date-to-now** would be classified as a short date.

> 5 minutes ago to 2019-10-09 23:59:59

aka **now-minus-offset-to-date** would be classified a long date, even though technically it's shorter than date-to-now in an example above.

With **isLongDate** both of them are classified as long dates.

Comment (davkal) [reply packages/grafana-ui/src/components/TimePicker/TimePicker.tsx]: I think for the purpose of collapsing the button, it's fine if either is already an absolute date, perhaps `isAbsolute` could change to `hasAbsolute` if that works with the rest of the code. 

I guess the biggest issue I have is that we're doing 2 different kinds of detection here: one with `typeof` and the other one with `isDateTime` just above. Can we use either and not both? If we need to use both, then I'd expect a comment.

Comment (Estrax) [reply packages/grafana-ui/src/components/TimePicker/TimePicker.tsx]: @davkal of course, we don't need them both. I already fixed that - renamed isAbsolute to hasAbsolute and consolidated long datetime detection within hasAbsolute.

```ts
const hasAbsolute = isDateTime(value.raw.from) || isDateTime(value.raw.to);
```

Comment (davkal) [reply packages/grafana-ui/src/components/TimePicker/TimePicker.tsx]: Happy with this.

**Comment by Estrax** [comment]
@davkal Fixed.

Normal view:
Screenshot from 2019-10-12 03-04-41

Split view:
Screenshot from 2019-10-12 03-04-26

Select open:
Screenshot from 2019-10-12 03-12-38

Requires a modified version of @torkelo/react-select, so it's going to be mergeable right when the fix is going to be published (very soon).

**Comment by aocenas** [comment]
@Estrax is the small version of the select going to be only in Explore? Because I assume this won't work with selects where values do not have image.

**Comment by Estrax** [comment]
@aocenas yes (unless you add `explore-toolbar` class somewhere else)

File: public/sass/pages/_explore.scss [hunk 2]
```diff
@@ -120,6 +128,33 @@
...
+  }
+
+  .explore-toolbar.splitted {
+    .explore-toolbar-content-item {
+      .datasource-picker {
+        .ds-picker {
+          min-width: 50px;
+          max-width: 50px;
+
+          .gf-form-select-box__single-value {
```
Comment (torkelo) [inline public/sass/pages/_explore.scss]: I am not sure I think overriding internal css classes from explore is ideal from a maintainable standpoint. Feels a bit fragile. 

Can this be done via prop option on the on the select component instead?

File: public/sass/pages/_explore.scss [hunk 3]
```diff
@@ -148,10 +183,29 @@
...
+  }
+}
+
+@media only screen and (max-width: 634px) {
+  .datasource-picker {
+    .ds-picker {
+      min-width: 50px;
+      max-width: 50px;
+
+      .gf-form-select-box__single-value {
```
Comment (torkelo) [inline public/sass/pages/_explore.scss]: This is an internal class, style defined in _Select.scss  , overriding it from the outside like this is not ok.

File: public/sass/pages/_explore.scss [hunk 4]
```diff
@@ -120,6 +128,33 @@
...
 }
 
+@media only screen and (max-width: 1210px) {
+  .time-picker-long-date {
+    .select-button {
+      span {
+        display: none;
+      }
+    }
+  }
```
Comment (torkelo) [inline public/sass/pages/_explore.scss]: The TimePicker is a generic component that is used from Dashboard (and potentially other places) as well, this should be in _TimePicker.scss

File: public/sass/pages/_explore.scss [hunk 5]
```diff
@@ -61,6 +61,14 @@
...
     margin-right: auto;
   }
+
+  .time-picker-long-date {
+    .select-button {
+      span {
+        display: none;
+      }
+    }
+  }
```
Comment (torkelo) [inline public/sass/pages/_explore.scss]: The TimePicker is a generic component that is used from Dashboard (and potentially other places) as well, this should be in _TimePicker.scss

**Review by torkelo — CHANGES_REQUESTED** [review]
This has some css challenges that might be tricky to address. Overriding internal css classes for generic components from the outside is very fragile and would like to find other options.

File: public/sass/pages/_explore.scss (hunk 2 above)
Comment (Estrax) [reply public/sass/pages/_explore.scss]: Not sure if the react-select has the prop indicating the split, but I can refactor it to ResponsiveButton.

File: public/sass/pages/_explore.scss (hunk 5 above)
Comment (Estrax) [reply public/sass/pages/_explore.scss]: Moved to _TimePicker.scss.

File: public/sass/pages/_explore.scss (hunk 4 above)
Comment (Estrax) [reply public/sass/pages/_explore.scss]: Moved to _TimePicker.scss.

File: public/sass/pages/_explore.scss (hunk 3 above)
Comment (Estrax) [reply public/sass/pages/_explore.scss]: Moved to _Select.scss and refactored it to be a prop.

**Comment by Estrax** [comment]
@torkelo moved these styles to appropriate files, rewrote the collapse and made it a prop. Is it a little bit better now?

**Review by torkelo — COMMENTED** [review]
Looks much better now, but still some things that do not feel right. 

the .datasource-picker class is a bit strange, should really be named .explore-ds-picker (and be defined in explore.css) as it not part of the data source picker. 

Then I think the ds-picker (the DataSourcePicker component) should have a class/property to make it with: 100% so you can control width from the outside, then the style rules on .datasource-picker does not need to override the internal class of DataSourcePicker. 

The collapse handling is also a bit off. It's about hiding the value, wondering if there is cleaner way to hide that value

**Review by torkelo — COMMENTED** [review]
Pushed 2 commits to show what I mean: 
Using a new prop on the value that hides the text (instead of isCollapsed on the component) 
Also using containerWidth from redux state
https://github.com/grafana/grafana/tree/hide-text-in-ds-picker 

It needs more work (especially the use of containerWidth) as it does not match the media queries for the width of the component

File: packages/grafana-ui/src/components/Select/Select.tsx [hunk 6]
```diff
@@ -50,6 +50,7 @@ export interface CommonProps<T> {
   onOpenMenu?: () => void;
   onCloseMenu?: () => void;
   tabSelectsValue?: boolean;
+  hideTextValue?: boolean;
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Select/Select.tsx]: this property on Select does not look like it is needed, can't see that it is used anywhere.

File: packages/grafana-ui/src/components/TimePicker/_TimePicker.scss [hunk 7]
```diff
@@ -250,6 +251,24 @@ $arrowPadding: $arrowPaddingToBorder * 3;
       }
     }
   }
+
+  .time-picker-long-date {
+    .select-button {
+      span {
```
Comment (torkelo) [inline packages/grafana-ui/src/components/TimePicker/_TimePicker.scss]: now we only have this internal override on an internal span element that belongs to the ButtonSelect component. So this time-picker-long-date class should be a property or BEM modifier class on the ButtonSelect not on the TimePicker component.

**Review by torkelo — COMMENTED** [review]
Great updates, we are getting close. Still have the TimePicker scss overriding an internal element type that belongs to the ButtonSelect component.

File: packages/grafana-ui/src/components/Select/Select.tsx (hunk 6 above)
Comment (Estrax) [reply packages/grafana-ui/src/components/Select/Select.tsx]: Right now it's being controlled by ExploreToolbar and passed down as a prop so it prevents hiding text from other Select components as they may use different breakpoints.

Comment (torkelo) [reply packages/grafana-ui/src/components/Select/Select.tsx]: but cannot see it being used anywhere, it's set on the SelectComponent (which does not have any such property)

File: packages/grafana-ui/src/components/Select/Select.tsx [hunk 8]
```diff
@@ -251,6 +256,7 @@ export class AsyncSelect<T> extends PureComponent<AsyncProps<T>> {
               }}
               defaultValue={defaultValue}
               value={value}
+              hideText={hideTextValue}
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Select/Select.tsx]: this is the one that I don't think is used, as there is no such property on this component.

File: packages/grafana-ui/src/components/Select/_Select.scss [hunk 9]
```diff
@@ -217,3 +220,7 @@ $select-input-bg-disabled: $input-bg-disabled;
 .select-button {
   display: flex;
 }
+
+.gf-form-select-box__img-value-hidden {
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Select/_Select.scss]: where is this class set?

File: packages/grafana-ui/src/components/Select/Select.tsx (hunk 6 above)
Comment (Estrax) [reply packages/grafana-ui/src/components/Select/Select.tsx]: Ah, you're absolutely right. My bad. I missed to remove it from there as I forwarded this prop via value object.

File: packages/grafana-ui/src/components/Select/_Select.scss (hunk 9 above)
Comment (Estrax) [reply packages/grafana-ui/src/components/Select/_Select.scss]: It was added when I was experimenting with pure CSS solution only. Did the same thing the other way, already cleaned it up.

**Review by torkelo — DISMISSED** [review]
Code wise this looks good now! Thanks keeping up with all my change requests! 

@davkal can you do final test?

**Comment by davkal** [comment]
Works well! Two last nits:

- in the non-split mode, the Explore is butted against the selector, would be great if the spaces between `Explore`, the selector, and the next button could be the same as the space between the rest of the toolbar buttons
Screenshot 2019-10-25 at 15 53 03

- datasources with long names look broken when they have mode switcher, e.g., Elastic, and the run button wraps to next line

Screenshot 2019-10-25 at 15 57 11

**Comment by Estrax** [comment]
> * in the non-split mode, the Explore is butted against the selector, would be great if the spaces between `Explore`, the selector, and the next button could be the same as the space between the rest of the toolbar buttons

Fixed.

> * datasources with long names look broken when they have mode switcher, e.g., Elastic, and the run button wraps to next line

Fixed - set up max-width for this picker and cut the displayed label for a currently selected option to up to 36 characters. Full labels available (with a tick indicating a selected source) within a dropdown when the picker is open.

**Comment by davkal** [comment]
Run button disappears on smaller screens (it's fine if it wraps on really small screens like this here):

Loki (live button stays for some reason):
Screenshot 2019-10-28 at 10 53 17

Elastic:
Screenshot 2019-10-28 at 10 53 40

**Comment by Estrax** [comment]
> Run button disappears on smaller screens (it's fine if it wraps on really small screens like this here):
> 
> Loki (live button stays for some reason):
> Screenshot 2019-10-28 at 10 53 17
> 
> Elastic:
> Screenshot 2019-10-28 at 10 53 40

Solved. There was a media query hiding refresh picker below the `sm` breakpoint, changed it to `xs`.

**Comment by Estrax** [comment]
Solved. There was a media query hiding refresh picker below the sm breakpoint, changed it to xs.

File: public/app/core/components/Select/DataSourcePicker.tsx [hunk 10]
```diff
@@ -45,10 +46,11 @@ export class DataSourcePicker extends PureComponent<Props> {
     }));
 
     const value = current && {
-      label: current.name,
+      label: current.name.substr(0, 37),
```
Comment (davkal) [inline public/app/core/components/Select/DataSourcePicker.tsx]: A bit harsh, but I like it.

**Review by davkal — COMMENTED** [review]
Works well and looks good!

Nits: A couple of magic numbers in there that could either be extracted or use comments to tell what the goal of those widths is.

**Review by davkal — APPROVED** [review]

**Pull request merged**